pkg/workload/pgx_helpers.go

```diff
@@ -319,32 +315,57 @@ func (m *MultiConnPool) WarmupConns(ctx context.Context, numConns int) error {
 	// (128).
 	g.SetLimit(100)
 
-	var warmupConnsPerPool []int
+	type warmupPool struct {
+		maxConns int
+		pool     *pgxpool.Pool
+	}
+	warmupPools := make([]warmupPool, len(m.Pools))
 	if numConns == 0 {
-		warmupConnsPerPool = make([]int, len(m.Pools))
 		for i, p := range m.Pools {
-			warmupConnsPerPool[i] = int(p.Config().MaxConns)
+			warmupPools[i].maxConns = int(p.Config().MaxConns)
+			warmupPools[i].pool = p
 		}
 	} else {
-		warmupConnsPerPool = distribute(numConns, len(m.Pools))
+		numPoolConns := distribute(numConns, len(m.Pools))
 		for i, p := range m.Pools {
-			poolMaxConns := int(p.Config().MaxConns)
-			if warmupConnsPerPool[i] > poolMaxConns {
-				warmupConnsPerPool[i] = poolMaxConns
+			warmupPools[i].pool = p
+			warmupPools[i].maxConns = numPoolConns[i]
+			maxPoolConns := int(p.Config().MaxConns)
+			if warmupPools[i].maxConns > maxPoolConns {
+				warmupPools[i].maxConns = maxPoolConns
 			}
 		}
 	}
 
 	var numWarmupConns int
-	for _, n := range warmupConnsPerPool {
-		numWarmupConns += n
+	for i, p := range warmupPools {
+		numWarmupConns += p.maxConns
+		if p.maxConns <= 0 {
+			warmupPools[i].maxConns = 1
+		}
 	}
+
+	// NOTE(seanc@): see context cancellation note below.
+	// TODO(seanc@): Change WithTimeout() back to WithCancel()
+	const maxWarmupTime = 5 * time.Minute // NOTE(seanc@): 5min == AWS NLB TCP idle time
+	const minWarmupTime = 15 * time.Second
+	const maxTimePerConn = 200 * time.Millisecond
+	warmupTime := minWarmupTime
+	switch {
+	case int(warmupTime) < numWarmupConns*int(maxTimePerConn):
+		warmupTime = time.Duration(numWarmupConns * int(maxTimePerConn))
+	case warmupTime > maxWarmupTime:
+		warmupTime = maxWarmupTime
+	}
+	warmupCtx, cancel := context.WithTimeout(ctx, warmupTime)
```

Comment:
nit: preference for just `ctx, cancel := ...` because that's the only context you'll be using, no need to keep the old one in scope.